server-timing/class-perflab-server-timing.php

```diff
@@ -0,0 +1,255 @@
+	 * @return bool True if registered, false otherwise.
+	 */
+	public function has_registered_metric( $metric_slug ) {
+		return isset( $this->registered_metrics[ $metric_slug ] ) && isset( $this->registered_metrics_data[ $metric_slug ] );
+	}
+
+	/**
+	 * Outputs the Server-Timing header.
+	 *
+	 * This method must be called before rendering the page.
+	 *
+	 * @since n.e.x.t
+	 */
+	public function send_header() {
+		if ( headers_sent() ) {
+			_doing_it_wrong(
+				__METHOD__,
+				__( 'The method must be called before headers have been sent.', 'performance-lab' ),
+				''
+			);
+			return;
+		}
+
+		/**
+		 * Fires right before the Server-Timing header is sent.
+		 *
+		 * This action is the last possible point to register a Server-Timing metric.
+		 *
+		 * @since n.e.x.t
+		 */
+		do_action( 'perflab_server_timing_send_header' );
+
+		$header_value = $this->get_header_value();
+		if ( ! $header_value ) {
+			return;
+		}
+
+		header( sprintf( 'Server-Timing: %s', $header_value ), false );
+	}
+
+	/**
+	 * Gets the value for the Server-Timing header.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return string The Server-Timing header value.
+	 */
+	public function get_header_value() {
+		// Get all metric header values, as long as the current user has access to the metric.
+		$metric_header_values = array_filter(
+			array_map(
+				function( Perflab_Server_Timing_Metric $metric ) {
+					$value = $metric->get_value();
+					if ( null === $value ) {
+						return null;
+					}
+
+					// Check the registered capability here to ensure no metric without access is exposed.
+					if ( ! current_user_can( $this->registered_metrics_data[ $metric->get_slug() ]['access_cap'] ) ) {
+						return null;
+					}
+
+					return $this->format_metric_header_value( $metric );
+				},
+				$this->registered_metrics
+			),
+			function( $value ) {
+				return null !== $value;
+			}
+		);
+
+		return implode( ', ', $metric_header_values );
+	}
+
+	/**
+	 * Returns whether an output buffer should be used to gather Server-Timing metrics during template rendering.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return bool True if an output buffer should be used, false otherwise.
+	 */
+	public function use_output_buffer() {
+		/**
+		 * Filters whether an output buffer should be used to be able to gather additional Server-Timing metrics.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @param bool $use_output_buffer Whether to use an output buffer.
+		 */
+		return apply_filters( 'perflab_server_timing_use_output_buffer', false );
+	}
+
+	/**
+	 * Hook callback for the 'template_include' filter.
+	 *
+	 * This effectively initializes the class to send the Server-Timing header at the right point.
+	 *
+	 * This method is solely intended for internal use within WordPress.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param mixed $passthrough Optional. Filter value. Default null.
+	 * @return mixed Unmodified value of $passthrough.
+	 */
+	public function on_template_include( $passthrough = null ) {
```

Comment:
core functions intended for internal use only are usually prefixed with `_`, maybe we should follow that here?

Reply:
@adamsilverstein This is actually mixed in core. There is no clear guidance. The last times I added `_` prefixed core functions, I always got pushback that they should simply have an `@access private` annotation and no longer use such a prefix.

Reply:
🤷🏼 OK!